qiskit/circuit/library/generalized_gates/linear_function.py

```diff
@@ -0,0 +1,174 @@
+
+    **Example:** the circuit
+
+    .. parsed-literal::
+
+        q_0: ──■──
+             ┌─┴─┐
+        q_1: ┤ X ├
+             └───┘
+        q_2: ─────
+
+    is represented by a 3x3 linear matrix
+
+    .. math::
+
+            \begin{pmatrix}
+                1 & 0 & 0 \\
+                1 & 1 & 0 \\
+                0 & 0 & 1
+            \end{pmatrix}
+
+
+    References:
+        [1]: Ketan N. Patel, Igor L. Markov, and John P. Hayes,
+             Optimal synthesis of linear reversible circuits,
+             Quantum Inf. Comput. 8(3) (2008).
+             `Online at umich.edu. <https://web.eecs.umich.edu/~imarkov/pubs/jour/qic08-cnot.pdf>`_
+    """
+
+    def __init__(
+        self,
+        linear: Union[List[List[int]], np.ndarray, QuantumCircuit],
+        validate_input: Optional[bool] = False,
+    ) -> None:
+        """Create a new linear function.
+
+        Args:
+            linear (list[list] or ndarray[bool] or QuantumCircuit):
+                either an n x n matrix, describing the linear function,
+                or a quantum circuit composed of linear gates only
+                (currently supported gates are CX and SWAP).
+
+            validate_input: if True, performs more expensive input validation checks,
+                such as checking that a given n x n matrix is invertible.
+
+        Raises:
+            CircuitError: if the input is invalid:
+                either a matrix is non {square, invertible},
+                or a quantum circuit contains non-linear gates.
+
+        """
+        if not isinstance(linear, (list, np.ndarray, QuantumCircuit)):
+            raise CircuitError(
+                "A linear function must be represented either by a list, "
+                "a numpy array, or a quantum circuit with linear gates."
+            )
+
+        if isinstance(linear, QuantumCircuit):
+            # The following function will raise a CircuitError if there are nonlinear gates.
+            self._linear = _linear_quantum_circuit_to_mat(linear)
+
+        else:
+            # Normalize to numpy array (coercing entries to 0s and 1s)
+            try:
+                self._linear = np.array(linear, dtype=bool, copy=True)
+            except ValueError:
+                raise CircuitError(
+                    "A linear function must be represented by a square matrix."
+                ) from None
+
+            # Check that the matrix is square
+            if not len(self._linear.shape) == 2 or self._linear.shape[0] != self._linear.shape[1]:
+                raise CircuitError("A linear function must be represented by a square matrix.")
+
+            # Optionally, check that the matrix is invertible
+            if validate_input:
+                det = np.linalg.det(linear) % 2
+                if not np.allclose(det, 1):
+                    raise CircuitError(
+                        "A linear function must be represented by an invertible matrix."
+                    )
```

Comment:
I'm slightly worried by the `np.linalg.det` check for scaling purposes here; the maximal value of the determinant for these binary matrices grows quickly with `n` (https://oeis.org/A003432), and there's clearly some precision loss because of the need for the fuzzy comparison - the answer _should_ be exact, but the (I'm assuming) LU factorisation introduces funniness.  We're probably ok to accept this for the time being, but it might be something we need to be aware of for the future.  We can probably write a custom routine for calculating determinants over the field {0, 1} if absolutely necessary.